sixtyfps_runtime/corelib/graphics/image.rs

```diff
@@ -54,39 +270,80 @@ impl Image {
+        Image(ImageInner::EmbeddedImage(SharedImageBuffer::RGB8(buffer)))
+    }
+
+    /// Creates a new Image from the specified shared pixel buffer, where each pixel has four color
+    /// channels (red, green, blue and alpha) encoded as u8.
+    pub fn from_rgba8(buffer: SharedPixelBuffer<RGBA8Pixel>) -> Self {
+        Image(ImageInner::EmbeddedImage(SharedImageBuffer::RGBA8(buffer)))
+    }
+
+    /// Creates a new Image from the specified shared pixel buffer, where each pixel has four color
+    /// channels (red, green, blue and alpha) encoded as u8 and, in contrast to [`Self::from_rgba8`],
+    /// the alpha channel is also assumed to be multiplied to the red, green and blue channels.
     ///
-    /// Returns an error if the `data` does not have a size of `width * height`.
-    pub fn load_from_argb(
-        width: u32,
-        height: u32,
-        data: SharedVector<u32>,
-    ) -> Result<Self, LoadImageError> {
-        if data.len() != width as usize * height as usize {
-            Err(LoadImageError(()))
-        } else {
-            Ok(Image(ImageReference::EmbeddedRgbaImage { width, height, data }))
-        }
+    /// Only construct an Image with this function if you know that your pixels are encoded this way.
+    pub fn from_rgba8_premultiplied(buffer: SharedPixelBuffer<RGBA8Pixel>) -> Self {
+        Image(ImageInner::EmbeddedImage(SharedImageBuffer::RGBA8Premultiplied(buffer)))
     }
-    */
 
     /// Returns the size of the Image in pixels.
     pub fn size(&self) -> crate::graphics::Size {
-        match crate::backend::instance() {
-            Some(backend) => backend.image_size(self),
-            None => panic!("sixtyfps::Image::size() called too early (before a graphics backend was chosen). You need to create a component first."),
+        match &self.0 {
+            ImageInner::None => Default::default(),
+            ImageInner::AbsoluteFilePath(_) |  ImageInner::EmbeddedData { .. } => {
+                match crate::backend::instance() {
+                    Some(backend) => backend.image_size(self),
+                    None => panic!("sixtyfps::Image::size() called too early (before a graphics backend was chosen). You need to create a component first."),
+                }
+            },
+            ImageInner::EmbeddedImage(buffer) => [buffer.width() as _, buffer.height() as _].into(),
         }
     }
 }
 
+#[test]
+fn test_image_size_from_buffer_without_backend() {
+    {
+        assert_eq!(Image::default().size(), Default::default());
+    }
+    {
+        let buffer = SharedPixelBuffer::<RGB8Pixel>::new(320, 200);
+        let image = Image::from_rgb8(buffer);
+        assert_eq!(image.size(), [320., 200.].into())
+    }
+}
+
 #[cfg(feature = "ffi")]
 pub(crate) mod ffi {
     #![allow(unsafe_code)]
 
     use super::super::Size;
     use super::*;
 
+    /// Expand RGB8Pixel so that cbindgen can see it. (is in fact rgb::RGB<u8>)
+    #[cfg(cbindgen)]
+    #[repr(C)]
+    struct RGB8Pixel {
```

Comment:
Alternatively, one could declare it in C++ directly in one of our internal headers

Reply:
True, and that might become the way to go if we make this image-from-data API public in C++.